<?xml version="1.0"?>
<robot>
  <gazebo>
    <!-- differential drive -->
    <plugin name="differential_drive_plugin_6w" filename="libdifferential_drive_plugin_6w.so">
      <legacyMode>false</legacyMode>
      <updateRate>50</updateRate>
      <midLeftJoint>center_left_wheel_joint</midLeftJoint>
      <midRightJoint>center_right_wheel_joint</midRightJoint>
      <frontLeftJoint>front_left_wheel_joint</frontLeftJoint>
      <frontRightJoint>front_right_wheel_joint</frontRightJoint>
      <rearLeftJoint>back_left_wheel_joint</rearLeftJoint>
      <rearRightJoint>back_right_wheel_joint</rearRightJoint>
      <wheelSeparation>${base_width}</wheelSeparation>
      <wheelDiameter>${wheel_radius * 2}</wheelDiameter>
      <torque>200</torque>
      <commandTopic>/cmd_vel</commandTopic>
      <bodyName>base_link</bodyName>
    </plugin>

    <!-- GPS -->
    <plugin name="gazebo_ros_gps" filename="libhector_gazebo_ros_gps.so">
      <updateRate>100.0</updateRate>
      <bodyName>base_link</bodyName>
      <frameId>base_link</frameId>
      <topicName>gps/fix</topicName>
      <velocityTopicName>gps/fix_velocity</velocityTopicName>
      <referenceLatitude>42.293195</referenceLatitude>
      <referenceLongitude>-83.7096706</referenceLongitude>
      <referenceHeading>90.0</referenceHeading>
      <referenceAltitude>0.0</referenceAltitude>
      <offset>0 0 0</offset>
      <drift>0.0 0.0 0.0</drift>
      <gaussianNoise>0.3 0.3 0.3</gaussianNoise>
      <velocityDrift>0.0 0.0 0.0</velocityDrift>
      <velocityGaussianNoise>0.2 0.2 0.2</velocityGaussianNoise>
    </plugin>
  </gazebo>

  <!-- IMU -->
  <gazebo reference="imu_link">
    <sensor name="imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>imu/imu_only</topicName>
        <bodyName>base_link</bodyName>
        <updateRateHZ>100.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <accelGaussianNoise>0.0 0.0 0.0</accelGaussianNoise>
        <accelDrift>0.0 0.0 0.0</accelDrift>
        <acccelDriftFrequency>0.0 0.0 0.0</acccelDriftFrequency>
        <accelOffset>0.0 0.0 0.0</accelOffset>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>base_link</frameName>
        <initialOrientationAsReference>false</initialOrientationAsReference>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
  </gazebo>

  <!-- magnetometer -->
  <gazebo>
    <plugin name="magnetometer_plugin" filename="libhector_gazebo_ros_magnetic.so">
      <updateRate>100</updateRate>
      <bodyName>base_link</bodyName>
      <topicName>imu/mag_only</topicName>
      <referenceHeading>90.0</referenceHeading>
      <declination>3.0333</declination>
      <inclination>60.0</inclination>
      <driftFrequency>0 0 0</driftFrequency>
      <gaussianNoise>0 0 0</gaussianNoise>
    </plugin>
  </gazebo>


  <!-- long range camera -->
  <gazebo reference="long_range_camera_link">
    <sensor type="camera" name="long_range_camera">
      <update_rate>5</update_rate>
      <camera>
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>1920</width>
          <height>1080</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
              That pixel's noise value is added to each of its color
              channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>long_range_cam</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>long_range_cam_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.00000001</distortionK1>
        <distortionK2>0.00000001</distortionK2>
        <distortionK3>0.00000001</distortionK3>
        <distortionT1>0.00000001</distortionT1>
        <distortionT2>0.00000001</distortionT2>
      </plugin>
    </sensor>
  </gazebo>

  <!-- depth camera -->
  <gazebo reference="camera_link">
    <sensor name="camera_link_camera" type="depth">
      <update_rate>15</update_rate>
      <camera>
        <horizontal_fov>1.919862</horizontal_fov>
        <image>
          <width>320</width>
          <height>240</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.2</near>
          <far>20</far>
        </clip>
      </camera>
      <plugin name="gazebo_ros_openni_kinect_plugin" filename="libkinect_plugin.so">
        <baseline>0</baseline> <!-- Publish same rate as camera -->
        <alwaysOn>true</alwaysOn>
        <updateRate>5</updateRate>
        <cameraName>camera/left</cameraName>
        <imageTopicName>image</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <depthImageTopicName>depth_image</depthImageTopicName>
        <depthImageInfoTopicName>camera_info</depthImageInfoTopicName>
        <pointCloudTopicName>points</pointCloudTopicName>
        <frameName>left_camera_link</frameName>
        <pointCloudCutoff>0.3</pointCloudCutoff>
        <pointCloudCutoffMax>20.0</pointCloudCutoffMax>
        <distortionK1>0.00000001</distortionK1>
        <distortionK2>0.00000001</distortionK2>
        <distortionK3>0.00000001</distortionK3>
        <distortionT1>0.00000001</distortionT1>
        <distortionT2>0.00000001</distortionT2>
        <CxPrime>0</CxPrime>
        <Cx>0</Cx>
        <Cy>0</Cy>
        <focalLength>0</focalLength>
        <hackBaseline>0</hackBaseline>
      </plugin>
    </sensor>
  </gazebo>
</robot>